### PR TITLE
fix(local-runtime): scope hybrid -ot spill to recurrent layers, reprice lean logits buffer

### LINKED ISSUE #113329: [Bug] Local runtime: 9B hybrid (qwen35) launched with FFN→CPU override on 8 GB despite fitting — overhead overestimated ~3.7 GiB, hybrid -ot moves every layer's FFN (2.4× slower gen)
### Bug Description

On an 8 GB card, the managed local runtime launches **Ornith-1.5-9B (arch `qwen35`, hybrid: 24 recurrent + 8 full-attention layers)** at the 64K floor with `--override-tensor blk\.\d+\.ffn_.*\.weight=CPU`, cutting generation speed ~2.4×. The model actually fits entirely on the GPU **within Hermes's own usable-VRAM budget**, so the spill decision is wrong, and when a spill *is* decided, the hybrid placement moves far more than needed.

Two separate problems in `hermes_cli/local_runtime/context_policy.py` (identical on `main` as of 4e9d3c713a):

**1. Launch footprint overestimated by ~3.7 GiB → false spill.** `plan_launch()` prices overhead as `RUNTIME_OVERHEAD_BYTES` (1.5 GiB) + `ub_logits_bytes(n_vocab)` = `2048 × n_vocab × 4` → 1.90 GiB for this model's 248,320 vocab. Hermes's decision for this machine:

| term | estimate |
|---|---|
| weights | 4.97 GiB |
| KV @ 64K (q8_0, FA) | 1.16 GiB |
| fixed overhead | 1.50 GiB |
| ub logits (n_ubatch 2048) | 1.90 GiB |
| **need** | **9.52 GiB** |
| usable budget (`probe_budget(planning=True)`, 8 GiB − 2 GiB margin) | 6.00 GiB |
| → `spill_bytes` | 3.53 GiB |

Measured with the exact launch args Hermes generates (`-c 65536 -ctk q8_0 -ctv q8_0 -fa on -b 2048 -ub 2048`) but no `-ot`: **5,960 MiB (5.82 GiB) resident, peaking at +30 MiB while processing a 15,912-token prompt** — so all non-weight costs combined were ~0.85 GiB vs. 4.55 GiB estimated, and the model fits under the 6.00 GiB budget with the full margin intact.

**2. Hybrid spill placement ignores both the layer kind and the spill size.** `spill_overrides()` says "hybrids push recurrent-layer FFNs", but for `profile.recurrent_layer_count > 0` it returns `blk\.\d+\.ffn_.*\.weight=CPU`, which matches the FFN tensors of **every** block, including the 8 full-attention layers. It's also all-or-nothing: the same pattern is emitted whether `spill_bytes` is 100 MB or 5 GB. (Pure dense models get `[]` and MoE the `_exps` pattern — this only affects hybrids.)

### Steps to Reproduce

1. Windows 11, RTX 3070 Laptop 8 GB, `local_runtime.enabled: true`, runtime `b10964` CUDA.
2. Stage a `qwen35` hybrid ~9B Q4 GGUF (here `Ornith-1.5-9B-uncensored.Q4_K_S.gguf`, 262K native ctx).
3. Start Hermes Desktop; inspect `runtimes/llamacpp/presets.ini` → section has `override-tensor = blk\.\d+\.ffn_.*\.weight=CPU`; `# hermes-decision` has `"spilled": true`.

Estimate breakdown reproduced with:

```python
from hermes_cli.local_runtime import presets
from hermes_cli.local_runtime.hardware import probe_budget
from hermes_cli.local_runtime.context_policy import plan_launch, spill_overrides
prof = presets.profile_from_gguf(presets.read_gguf_header(path))
plan = plan_launch(prof, probe_budget(planning=True))
# moe=False, recurrent_layer_count=24, weights 4.97 GiB, n_vocab 248320
# WindowDecision(window=65536, spill_bytes=3784574976, ...), overhead 3.39 GiB
# spill_overrides(prof) == ['-ot', 'blk\\.\\d+\\.ffn_.*\\.weight=CPU']
```

### Expected Behavior

Model launches fully on GPU (no `-ot`), since measured resident + peak fits the usable budget. If a genuine spill is needed for a hybrid, offload only what the spill requires (and only recurrent-layer FFNs, per the docstring).

### Actual Behavior

A/B on a standalone `llama-server` with the generated preset args, same prompts, temperature 0, nothing else on the GPU:

| config | VRAM (model) | gen, 256-tok answer | 15,912-tok prompt: prefill | 15,912-tok prompt: gen |
|---|---|---|---|---|
| Hermes preset (`-ot` FFN→CPU) | 4,303 MiB | **8.1 tok/s** | 1,421.9 tok/s | **8.6 tok/s** |
| same, no `-ot` | 5,960 MiB | **20.8 tok/s** | 1,607.9 tok/s | **19.2 tok/s** |

The override saves 1.66 GiB of VRAM that wasn't needed and costs 2.2–2.6× generation speed. Because `presets.ini` is regenerated on every cold boot, hand-editing it doesn't stick (same pain point as #106895 §2).

### Affected Component

Local runtime (managed llama.cpp) — `context_policy.py` (`plan_launch`, `ub_logits_bytes`, `spill_overrides`); seen via Hermes Desktop.

### Operating System

Windows 11 Pro (26200), RTX 3070 Laptop 8 GB (WDDM), Ryzen 7 5800H, 16 GB RAM.

### Python Version

3.11.16 (bundled runtime)

### Hermes Version

0.21.3 (3c3ab69abb); llama.cpp runtime b10964 CUDA

### Root Cause Analysis

- `ub_logits_bytes` prices a full `n_ubatch × n_vocab` float32 logits buffer on the GPU. Its docstring says it was calibrated on Qwen3.8 Q4; on this 248K-vocab model the measured total of KV + compute buffers + logits was ~0.85 GiB, so the 1.90 GiB logits term alone exceeds reality, and it stacks on the 1.5 GiB fixed constant. The overestimate grows with vocab size, so large-vocab models on small cards are the ones pushed into false spills.
- `spill_overrides` for hybrids uses a block-agnostic regex and ignores `spill_bytes`.

### Proposed Fix

Suggestions (no patch, since the calibration data isn't mine to change blind):

1. Recalibrate the lean-posture overhead (`RUNTIME_OVERHEAD_BYTES` + `ub_logits_bytes` for `mtp_capable=False`) against large-vocab models, or measure-and-cache the real resident size after the first successful load and use it for subsequent preset generation.
2. In `spill_overrides`, build the hybrid pattern from the recurrent layer indices only (e.g. `blk\.(0|1|2|4|…)\.ffn_.*\.weight=CPU`), as the docstring describes.
3. Size the placement to `spill_bytes`: offload only enough FFN blocks to cover the spill, rather than all of them for any non-zero spill.
4. Related to #106895: a persistent per-model override (e.g. "never spill" / `extra_args`) that survives preset regeneration would let users opt out when they've verified a fit.

Current local workaround: an external watcher that strips the `override-tensor` line from `presets.ini` and restarts the idle router, which Hermes's supervisor respawns from the edited file.

### PR DESCRIPTION
**Pull request by kvnloo**
## What does this PR do?

Fixes two overestimation bugs in the local-runtime context policy that pushed a 9B hybrid model (Ornith-1.5-9B, qwen35, 24 recurrent + 8 full-attention layers) into an unnecessary FFN→CPU spill on an 8 GB card, costing ~2.4x generation speed.

1. **Hybrid `-ot` placement ignored layer kind.** `spill_overrides()` documented "hybrids push recurrent-layer FFNs" but returned `blk\.\d+\.ffn_.*\.weight=CPU`, matching the FFN tensors of *every* block including the full-attention layers. It now builds the pattern from the recurrent layer indices only (e.g. `blk\.(16|17|…)\.ffn_.*\.weight=CPU`), derived from `ModelProfile.layers` which is already in `blk.N` order.

2. **Lean-posture logits overhead overestimated ~3.7 GiB.** `ub_logits_bytes()` priced a full `n_ubatch × n_vocab × 4` float32 logits buffer on the GPU in the non-MTP path (1.9 GiB at 248K vocab). The issue's measured resident shows no such buffer materializes in the lean path (~0.85 GiB total non-weight overhead). The lean posture now prices the decode microbatch (`512 × vocab × 4`), consistent with the MTP decode posture's per-microbatch term.

The remaining gap to the issue's expected behavior (fully resident launch on the reporter's 8 GB card) is the 1.5 GiB `RUNTIME_OVERHEAD_BYTES` constant, which is grounded in a 32 GiB-card measurement — recalibrating it needs multi-hardware data, deliberately left for a follow-up rather than changed blind from one data point.

## Related Issue

Fixes #113329

## Type of Change

- [x] 🐛 Bug fix (non-breaking change that fixes an issue)

## Changes Made

- `hermes_cli/local_runtime/context_policy.py`: `spill_overrides()` names recurrent indices; `ub_logits_bytes()` lean posture repriced to decode microbatch; `LayerKind` import added
- `tests/hermes_cli/test_context_policy.py`: new invariant `test_hybrid_spill_placement_names_only_recurrent_layers` (proven red on base); updated `test_ub_logits_bytes_prices_the_flag_choice`

## How to Test

1. `scripts/run_tests.sh tests/hermes_cli/test_context_policy.py` — 32 passed
2. Neighboring suites: `test_local_preset_admission.py`, `test_local_growth.py`, `test_local_models_routes.py`, `test_catalog_variants.py` (42 passed), `test_local_runtime.py` (55 passed)
3. On the reporter's config: `plan_launch()` overhead for the 248K-vocab hybrid drops ~1.4 GiB and `spill_overrides()` emits only the recurrent indices

### Code

- [x] My commit messages follow Conventional Commits
- [x] I searched for existing PRs (open-PR search + issue timeline for "113329" — no contention)
- [x] My PR contains only changes related to this fix
- [x] I've run the test suites via `scripts/run_tests.sh` and all pass
- [x] I've added tests for my changes (1 new invariant test, red-on-base proven)
- [x] I've tested on my platform: Linux (sandbox, no GPU — unit-level only)

### Documentation & Housekeeping

- [x] Docstrings updated (`ub_logits_bytes`, `spill_overrides` comment)
- [x] No config keys changed — N/A
- [x] Cross-platform: pure-Python policy code, no platform branching — N/A

---
authored with AI assistance (Muse, Meta's Muse Spark) under the contributor's direction; the contributor reviewed and takes responsibility for the changes.

### CODE REVIEW

**Comment by Godhet** [comment]
Potential accounting mismatch in the lean non-MTP path: `launch_args()` still emits `-b 2048 -ub 2048` for `mtp_capable=False`, while `ub_logits_bytes()` now charges only `512 * vocab * 4`. The function doc says its price matches the microbatch posture chosen by `launch_args`, but these values no longer agree. Unless llama.cpp demonstrably avoids materializing the logits buffer for the 2048 decode microbatch (and that is separately accounted for), `plan_launch()` can under-estimate VRAM by 4x for non-MTP models and choose a zero-spill/rung that does not fit. Please add a production-path receipt/test or adjust the estimator/launch flag so the accounting contract is explicit.

**Comment by kvnloo** [comment]
Superseded by narrower current-main draft #120778 for the recurrent-layer placement bug. I am dropping this PR's logits-estimator change after @Godhet's review: current `launch_args()` still emits non-MTP `-ub 2048`, and the existing accounting test pins that posture, so charging a 512-token term without a production allocation receipt would make the estimator contract internally inconsistent. #120778 keeps only the mechanically provable fix: hybrid `-ot` may match recurrent FFNs, not full-attention FFNs. Closing this mixed-scope carrier; the estimator/calibration question remains open in #113329.
